webrtc/README.md

```diff
@@ -0,0 +1,323 @@
+  the remote telling the local node what port it is perceived on.
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+  Yes. For the lifetime of the page, one can generate a certificate once and
+  reuse it across connections. See also
+  https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection#using_certificates
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
+  based on it?
+
+## Connection Security
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel and starts a Noise handshake using _A_'s and
+   _B_'s libp2p identity. See
+   [noise-libp2p](https://github.com/libp2p/specs/tree/master/noise).
+
+2. _A_ and _B_ write their TLS certificate fingerprint on the negotiated Noise
+   channel as multibase encoded multihashes (see [#addressing]).
+
+3. _A_ and _B_ read the other sides TLS certificate fingerprint on the
+   negotiated Noise channel and compare it to the ones verified during the DTLS
+   handshake.
+
+   Note: WebRTC supports different hash functions to hash the TLS certificate
+   (see https://datatracker.ietf.org/doc/html/rfc8122#section-5). The hash
+   function used in WebRTC, the hash function used for the hash exchanged in the
+   additional Noise handshake and the hash function used in the multiaddr
+   `/certhash` component MUST be the same. On mismatch the final Noise handshake
+   MUST fail.
+
+### Open Questions
+
+- Can a _Browser_ access the fingerprint of its TLS certificate?
+
+  Chrome allows you to access the fingerprint of any locally-created certificate
+  directly via `RTCCertificate#getFingerprints`. Firefox does not allow you to
+  do so. Browser compatibility can be found
+  [here](https://developer.mozilla.org/en-US/docs/Web/API/RTCCertificate). In
+  practice, this is not an issue since the fingerprint is embedded in the local
+  SDP string.
+
+- Is the above proposed additional handshake secure? See also alternative
+  proposed Handshake for
+  [WebTransport](https://github.com/libp2p/specs/pull/404).
+
+- Would it be more efficient for _B_ to initiate the Noise handshake? In other
+  words, who is able to write on an established WebRTC connection first? _A_ or
+  _B_?
+
+- Instead of exchanging the TLS fingerprints once the Noise handshake finished,
+  could one instead attach the fingerprints to the Noise handshake messages as
+  additional payloads? That would reduce the overall connection establishment
+  latency by one round trip. Would this schema be secure using the Noise XX
+  handshake pattern?
```

Comment:
I would very much appreciate input on the above from folks more familiar with the Noise handshake framework.

@Kubuxu would you mind taking a look at this proposal?

@Menduist this is the optimization I mentioned in the libp2p community call. Would you mind taking a look and potentially forwarding this internally to the "person that speaks Noise" at status.im?

Related comment by @marten-seemann: https://github.com/libp2p/specs/pull/412/files#r922796816

Reply:
Had a talk with @s1fr0

Since you can retrieve with confidence the local & remote TLS fingerprint, one option to shave the RTT (and have a overall cleaner structure) is to put both fingerprints in the [prologue](https://noiseprotocol.org/noise.html#prologue) of noise.

The prologue is an arbitrary byte sequence that is not shared during the Noise handshake, but has to match on both peers, otherwise the handshake **will** fail
So you could simply use `(client_fingerprint || server_fingerprint)` as a prologue, do the handshake, and as soon as it succeeds, you are safe. The full auth procedure is thus 1 RTT

Another option would be to include it someway in the PSK.

Reply:
Yes, we can attach the fingerprints as additional payloads in the Noise XX handshake.
The initiator would attach their fingerprint to the first message, and the responder to the second message.

It is based on [Section 2.2 of Noise](http://noiseprotocol.org/noise.html#overview-of-handshake-state-machine (edited) ) spec:
> A handshake message consists of some DH public keys followed by a payload. The payload may contain certificates or other data chosen by the application. To send a handshake message, the sender specifies the payload and sequentially processes each token from a message pattern. [...]

> After processing the final token in a handshake message, the sender then writes the payload into the message buffer, encrypting it if k is non-empty, and hashes the output along with the old h to derive a new h.

> Encryption with k uses some AEAD cipher mode (in the sense of Rogaway [1]) and uses the current h value as associated data which is covered by the AEAD authentication.

If during decryption, the h is different from expected, the decryption will fail.

Reply:
> Since you can retrieve with confidence the local & remote TLS fingerprint, one option to shave the RTT (and have a overall cleaner structure) is to put both fingerprints in the prologue of noise.

Putting it into the Prologue is an even cleaner idea. I like it.

Reply:
Thank you @Menduist, @s1fr0 and @Kubuxu. This is very valuable input!

I suggest we leverage the Noise Prologue mechanism then, instead of exchanging the fingerprints after the handshake succeeded. @melekes, @ckousik, @marten-seemann any objections? If not, I would document this in detail here in the specification, which might also make the implementation work easier.

Reply:
> @melekes, @ckousik, @marten-seemann any objections?

none from me. I'll just need a Rust API (if it doesn't exist already).

Reply:
Documented Noise Prologue mechanism in ae1020b.

> So you could simply use (client_fingerprint || server_fingerprint) as a prologue

Is the plain concatenation of the two fingerprints enough, or do we need a context identifier (e.g. `"webrtc-noise-handshake:<MULTIHASH_A_FINGERPRINT><MULTIHASH_B_FINGERPRINT>"`) to prevent any signatures or hashes to be misused in other contexts @Menduist, @s1fr0 and @Kubuxu?

Reply:
> > any objections?
> 
> none from me. I'll just need a Rust API (if it doesn't exist already).

@melekes `snow` (Rust Noise crate) supports _Prologue_, though we do not expose it in `libp2p-noise` yet. Tracking this here https://github.com/libp2p/rust-libp2p/issues/2791.

Reply:
> Is the plain concatenation of the two fingerprints enough, or do we need a context identifier (e.g. "webrtc-noise-handshake:<MULTIHASH_A_FINGERPRINT><MULTIHASH_B_FINGERPRINT>") to prevent any signatures or hashes to be misused in other contexts @Menduist, @s1fr0 and @Kubuxu?

If I understand correctly, this is only the input to a hash function, but we should put a label in front nevertheless.

Reply:
> From what I understood, at this point there is a strong context already defined, since the handshake happens in a TLS connection where such fingerprints are already mutually authenticated.

At the point before noise handshake takes place, the browser has authenticated the server's TLS certificate but the server has not authenticated the browsers TLS certificate (there is no secure sdp exchange so we cannot exchange fingerprints in advance). We are relying on the noise handshake to authenticate using the peer Id. The assumption here is that the browser knows the correct peer id for the server.

Reply:
Thanks for the input! https://github.com/libp2p/specs/pull/412/commits/4b8e890c49061eb62c9d16a9afd3aa0a738e6285 adds the signature tag and requires the two fingerprints to be sorted, instead of requiring the fingerprint of the connection initiator to go first.

The latter (order of fingerprints) might prove useful in the cases where it is not clear who is the initiator, e.g. two browsers whole punching _at the same time_. We can later on specify how to decide who opens the datachannel. One option would be differentiating by the smaller fingerprint.

Reply:
Going to _resolve_ this conversation. Thanks for the input everyone.